students/javarb/sessions/2019/may/Session_34-35-Serialization/Session_34-35-Serialization.md

```diff
@@ -0,0 +1,225 @@
+- Stringtree.
+- SOJO.
+- json-taglib.
+- Flexjson.
+- Argo.
+- jsonij.
+- fastjson.
+- mjson.
+- jjson.
+- json-simple.
+- json-io.
+- google-gson.
+- FOSS Nova JSON.
+- Corn CONVERTER.
+- Apache johnzon.
+- Genson
+- cookjson.
+- progbase.
+
+Stil is a extensive list it's strange Jackson isn't there.
+
+## Session 34 (21/05/2019 )
+
+#### Part 2 - Benchmark project
+
+I decided to use: **Gson, Jackson and FastJson ** for my checks. In order to use them I needed to include them on `build.gradle` file:
+```java
+dependencies {
+	//...
+	compile group: 'com.google.code.gson', name: 'gson', version: '2.8.5'
+	compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.9.8'
+	compile group: 'com.alibaba', name: 'fastjson', version: '1.2.56'
+}
+```
+
+In my first approximation  solution. I was using a single class called `Serialization.java` and a POJO called `Person.java`. 
+The idea behind the benchmark was use a single time in miliseconds comparison after serialize a list set of persons. 
+
+Receiving feedback of Roger, I finished to write `main()` method of `Serialization.java`. There, a for loop was populating a `List` of `Person` objects and passing over another for loop to store the benchmark results of calculate the miliseconds time that takes to serialize the list of `Person`s. That results are stored in another `List` that is passes to `mean` method in order to calculate mean value and print it:
+```java
+//...
+List<Person> persons = new ArrayList<>();
+
+for (int i = 0; i < 20000; i++) {
+	Person person = new Person();
+    person.setId(i);
+    person.setName("Name_" + i);
+    person.setAge(i % 100);
+    persons.add(i, person);
+}
+
+List<Long> gsonResults = new ArrayList<>();
+List<Long> jacksonResults = new ArrayList<>();
+List<Long> fastjsonResults = new ArrayList<>();
+
+for (int i = 0; i < 150; i++) {
+	gsonResults.add(benchmarkGson(persons));
+	jacksonResults.add(benchmarkJackson(persons));
+	fastjsonResults.add(benchmarkFastjson(persons));
+}
+
+long gsonMean = mean(gsonResults);
+System.out.println("Gson = " + gsonMean + " ms");
+
+long jacksonMean = mean(jacksonResults);
+System.out.println("Gson = " + gsonMean + " ms");
+
+long fastjsonMean = mean(fastjsonResults);
+System.out.println("Gson = " + gsonMean + " ms");
+//...
+```
+
+Following code chunk, performs serialization returning miliseconds time benchmark in this example for Gson serialization library:
+```java
+private static long benchmarkGson(List<Person> persons) {
+	long begin = System.currentTimeMillis();
+	new Gson().toJson(persons);
+	return System.currentTimeMillis() - begin;
+}
+```
+
+Finally, method on charge of return the mean can be seen following:
+```java
+private long mean(List<Long> benchmarks) {
+	long sum = 0;  
+	for (Long benchmark : benchmarks) {
+		sum += benchmark;
+	}
+	return sum / benchmarks.size();
+}
+```
+
+A for loop was used in order to add up benchmarks in order to after get the final mean time benchmarked. 
+
```

Comment:
With regards to benchmarking - actually the median might be better to use than the mean. These notes come from something I was reading yesterday, and I include them because it's interesting, and it might be useful for you to write why you use the mean as a benchmark, and that there are other options out there:

- The mean is not a very good metric if you want to know your “typical” response time, because it doesn’t tell you how many users actually experienced that delay.
- Usually it is better to use percentiles. If you take your list of response times and sort it from fastest to slowest, then the median is the halfway point: for example, if your median response time is 200 ms, that means half your requests return in less than 200 ms, and half your requests take longer than that.
- This makes the median a good metric if you want to know how long users typically have to wait: half of user requests are served in less than the median response time, and the other half take longer than the median. 
- The median is also known as the 50th percentile, and sometimes abbreviated as p50. 
- Note that the median refers to a single request; if the user makes several requests (over the course of a session, or because several resources are included in a single page), the probability that at least one of them is slower than the median is much greater than 50%.
- In order to figure out how bad your outliers are, you can look at higher percentiles: the 95th, 99th, and 99.9th percentiles are common (abbreviated p95, p99, and p999).
- High percentiles of response times, also known as tail latencies, are important because they directly affect users’ experience of the service. Reducing response times at very high percentiles is difficult because they are easily affected by random events outside of your control, and the benefits are diminishing.
- Percentiles are often used in service level objectives (SLOs) and service level agreements (SLAs), contracts that define the expected performance and availability of a service. These metrics set expectations for clients of the ser‐ vice and allow customers to demand a refund if the SLA is not met.

It also goes on to say:
> Beware that averaging percentiles, e.g., to reduce the time resolution or to com‐ bine data from several machines, is mathematically meaningless—the right way of aggregating response time data is to add the histograms.

- from Martin Kleppmann - Designing Data-Intensive Applications: The Big Ideas Behind Reliable, Scalable, and Maintainable Systems (2017, O’Reilly Media) pp. 14-16.

Reply:
You're absolutely right, and in fact we should be measuring more metrics than just mean. (E.g. 99th percentile, median, standard deviation etc.). But for the purposes of this exercise, the mean alone is good enough :)

Reply:
I think this information is pretty much useful I will check on this.